staging/vhost-device-gpu/src/device.rs

```diff
@@ -1385,4 +1390,227 @@ mod tests {
         backend.handle_event(0, EventSet::IN, &[vring], 0).unwrap();
     }
     }
+
+    mod test_image {
+        use super::*;
+        const GREEN_PIXEL: u32 = 0x00FF00FF;
+        const RED_PIXEL: u32 = 0xFF0000FF;
+        const BYTES_PER_PIXEL: usize = 4;
+
+        pub fn write(mem: &GuestMemoryMmap, image_addr: GuestAddress, width: u32, height: u32) {
+            let mut image_addr: u64 = image_addr.0;
+            for i in 0..width * height {
+                let pixel = if i % 2 == 0 { RED_PIXEL } else { GREEN_PIXEL };
+                let pixel = pixel.as_slice();
+
+                mem.memory()
+                    .write_slice(pixel, GuestAddress(image_addr))
+                    .unwrap();
+                image_addr += BYTES_PER_PIXEL as u64;
+            }
+        }
+
+        pub fn assert(data: &[u8], width: u32, height: u32) {
+            assert_eq!(data.len(), (width * height) as usize * BYTES_PER_PIXEL);
+            for (i, pixel) in data.chunks(BYTES_PER_PIXEL).enumerate() {
+                let expected_pixel = if i % 2 == 0 { RED_PIXEL } else { GREEN_PIXEL };
+                assert_eq!(
+                    pixel,
+                    expected_pixel.to_ne_bytes(),
+                    "Wrong pixel at index {i}"
+                );
+            }
+        }
+    }
+
+    fn split_into_mem_entries(
+        addr: GuestAddress,
+        len: u32,
+        chunk_size: u32,
+    ) -> Vec<virtio_gpu_mem_entry> {
+        let mut entries = Vec::new();
+        let mut addr = addr.0;
+        let mut remaining = len;
+
+        while remaining >= chunk_size {
+            entries.push(virtio_gpu_mem_entry {
+                addr,
+                length: len,
```

Comment:
Yes it should be. Good catch!

Interestingly the test should have failed, but it didn't, because I forgot to join this thread here:
```
thread::spawn(move || {
    let mut scanout_request_hdr = [0; 12];
    let mut scanout_request = VhostUserGpuScanout::default();
    let mut update_request_hdr = [0; 12];
    let mut update_request = VhostUserGpuUpdate::default();
    let mut result_img = vec![0xdd; (IMAGE_WIDTH * IMAGE_HEIGHT * 4) as usize];

    gpu_frontend.read_exact(&mut scanout_request_hdr).unwrap();
    gpu_frontend
        .read_exact(scanout_request.as_mut_slice())
        .unwrap();
    gpu_frontend.read_exact(&mut update_request_hdr).unwrap();
    gpu_frontend
        .read_exact(update_request.as_mut_slice())
        .unwrap();
    gpu_frontend.read_exact(&mut result_img).unwrap();

    assert_eq!(scanout_request, EXPECTED_SCANOUT_REQUEST);
    assert_eq!(update_request, EXPECTED_UPDATE_REQUEST);
    test_image::assert(&result_img, IMAGE_WIDTH, IMAGE_HEIGHT);
})
```
Since the thread wasn't joined this was actually a race condition where the asserts didn't finish running... So I am fixing this too.